plugin/auth_pam/auth_pam.c

```diff
@@ -150,6 +154,37 @@ static int pam_auth(MYSQL_PLUGIN_VIO *vio, MYSQL_SERVER_AUTH_INFO *info)
   PAM_DEBUG((stderr, "PAM: pam_start(%s, %s)\n", service, info->user_name));
   DO( pam_start(service, info->user_name, &pam_start_arg, &pamh) );
 
+  vio->info(vio, &vio_info);
+  if (vio_info.protocol == MYSQL_VIO_TCP)
+  {
+    int err_code;
+    char ip_str[INET6_ADDRSTRLEN];
+    const char *ip_str_ptr;
+    struct sockaddr_storage addr;
+    socklen_t addr_length= sizeof (addr);
+    /* Get sockaddr by socked fd. */
+
+    err_code= getpeername(vio_info.socket,  (struct sockaddr*) &addr, &addr_length);
+
+    if (err_code)
+    {
+      PAM_DEBUG((stderr, "getpeername() gave error: %d", errno));
+    }
+    else
+    {
+      struct sockaddr_in *s4= (struct sockaddr_in *) &addr;
+      struct sockaddr_in6 *s6= (struct sockaddr_in6 *) &addr;
+      ip_str_ptr= inet_ntop(addr.ss_family,
+                            addr.ss_family==AF_INET ? (const void *) &s4->sin_addr : (const void *) &s6->sin6_addr,
+                            ip_str,
+                            sizeof(ip_str));
+      if (ip_str_ptr)
+      {
+        PAM_DEBUG((stderr, "PAM: pam_set_item(PAM_RHOST, %s)\n", ip_str_ptr));
+        DO( pam_set_item(pamh, PAM_RHOST, ip_str_ptr) );
+      }
```

Comment:
Why do you want to set `PAM_RHOST`? What's the use case?

Reply:
PAM_RHOST provides information useful in diagnosis if there is an authentication failure. Like pam_tally2 (https://www.tecmint.com/use-pam_tally2-to-lock-and-unlock-ssh-failed-login-attempts/) can show you which host has the wrong authentication set.

Modules like pam_ihosts.so an restrict access based on IP.

With host information in the log, tools like fail2ban (https://github.com/fail2ban/fail2ban/blob/master/config/filter.d/pam-generic.conf) can block repeated failure attempts by IP.